airflow-core/src/airflow/ui/src/pages/XCom/XComFilters.tsx

```diff
@@ -16,206 +16,151 @@
+        label: translate("admin:columns.key"),
+        placeholder: translate("common:filters.keyPlaceholder"),
+        type: "text",
+      },
+      {
+        icon: <MdDateRange />,
+        key: SearchParamsKeys.LOGICAL_DATE_GTE,
+        label: translate("common:filters.logicalDateFromPlaceholder"),
+        placeholder: translate("common:filters.logicalDateFromPlaceholder"),
+        type: "date",
+      },
+      {
+        icon: <MdDateRange />,
+        key: SearchParamsKeys.LOGICAL_DATE_LTE,
+        label: translate("common:filters.logicalDateToPlaceholder"),
+        placeholder: translate("common:filters.logicalDateToPlaceholder"),
+        type: "date",
+      },
+      {
+        icon: <MdDateRange />,
+        key: SearchParamsKeys.RUN_AFTER_GTE,
+        label: translate("common:filters.runAfterFromPlaceholder"),
+        placeholder: translate("common:filters.runAfterFromPlaceholder"),
+        type: "date",
+      },
+      {
+        icon: <MdDateRange />,
+        key: SearchParamsKeys.RUN_AFTER_LTE,
+        label: translate("common:filters.runAfterToPlaceholder"),
+        placeholder: translate("common:filters.runAfterToPlaceholder"),
+        type: "date",
+      },
+    ];
+
+    if (dagId === "~") {
+      configs.push({
+        icon: <DagIcon />,
+        key: SearchParamsKeys.DAG_DISPLAY_NAME_PATTERN,
+        label: translate("common:dagName"),
+        placeholder: translate("common:filters.dagDisplayNamePlaceholder"),
+        type: "text",
+      });
+    }
+
+    if (runId === "~") {
+      configs.push({
+        icon: <FiBarChart />,
+        key: SearchParamsKeys.RUN_ID_PATTERN,
+        label: translate("common:runId"),
+        placeholder: translate("common:filters.runIdPlaceholder"),
+        type: "text",
+      });
+    }
+
+    if (taskId === "~") {
+      configs.push({
+        icon: <TaskIcon />,
+        key: SearchParamsKeys.TASK_ID_PATTERN,
+        label: translate("common:taskId"),
+        placeholder: translate("common:filters.taskIdPlaceholder"),
+        type: "text",
+      });
+    }
+
+    if (mapIndex === "-1") {
+      configs.push({
+        icon: <MdNumbers />,
+        key: SearchParamsKeys.MAP_INDEX,
+        label: translate("common:mapIndex"),
+        min: -1,
+        placeholder: translate("common:filters.mapIndexPlaceholder"),
+        type: "number",
+      });
+    }
+
+    return configs;
+  }, [dagId, mapIndex, runId, taskId, translate]);
+
+  const initialValues = useMemo(() => {
+    const values: Record<string, FilterValue> = {};
 
-  const handleFilterChange = useCallback(
-    (paramKey: string) => (value: string) => {
-      if (value === "") {
-        searchParams.delete(paramKey);
-      } else {
-        searchParams.set(paramKey, value);
+    filterConfigs.forEach((config) => {
+      const value = searchParams.get(config.key);
+
+      if (value !== null && value !== "") {
+        values[config.key] = config.type === "number" ? Number(value) : value;
```

Comment:
```suggestion
        const parsedValue = Number(value);
        values[config.key] = config.type === "number" && !isNaN(parsedValue) ? parsed : value;
```

It could be more type-safe in this way since

```js
Number("a") -> NaN
```

Reply:
Sure that's a great point 💪